server/expression/in_subquery.go

```diff
@@ -0,0 +1,259 @@
+	pgtypes "github.com/dolthub/doltgresql/server/types"
+)
+
+// InSubquery represents a VALUE IN (SELECT ...) expression.
+type InSubquery struct {
+	leftExpr  sql.Expression
+	rightExpr *plan.Subquery
+
+	// These variables are used so that we can resolve the comparison functions once and reuse them as we iterate over rows.
+	// These are assigned in WithChildren, so refer there for more information.
+	leftLiteral   *Literal
+	rightLiterals []*Literal
+	compFuncs     []*framework.CompiledFunction
+}
+
+var _ vitess.Injectable = (*InSubquery)(nil)
+var _ sql.Expression = (*InSubquery)(nil)
+var _ expression.BinaryExpression = (*InSubquery)(nil)
+
+// NewInSubquery returns a new *InSubquery.
+func NewInSubquery() *InSubquery {
+	return &InSubquery{}
+}
+
+// Children implements the sql.Expression interface.
+func (in *InSubquery) Children() []sql.Expression {
+	return []sql.Expression{in.leftExpr, in.rightExpr}
+}
+
+var nilKey, _ = sql.HashOf(sql.NewRow(nil))
+
+// Eval implements the sql.Expression interface.
+func (in *InSubquery) Eval(ctx *sql.Context, row sql.Row) (any, error) {
+	if len(in.compFuncs) == 0 {
+		return nil, fmt.Errorf("%T: cannot Eval as it has not been fully resolved", in)
+	}
+
+	left, err := in.leftExpr.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	// The NULL handling for IN expressions is tricky. According to
+	// https://dev.mysql.com/doc/refman/8.0/en/comparison-operators.html#operator_in:
+	// To comply with the SQL standard, IN() returns NULL not only if the expression on the left hand side is NULL, but
+	// also if no match is found in the list and one of the expressions in the list is NULL.
+	// However, there's a strange edge case. NULL IN (empty list) return 0, not NULL.
+	leftNull := left == nil
+
+	if types.NumColumns(in.Left().Type()) != types.NumColumns(in.Right().Type()) {
+		return nil, sql.ErrInvalidOperandColumns.New(types.NumColumns(in.Left().Type()), types.NumColumns(in.Right().Type()))
+	}
+
+	typ := in.rightExpr.Type()
+	right := in.rightExpr
+
+	// TODO: does this work for all pg values?
+	values, err := right.HashMultiple(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	// NULL IN (list) returns NULL. NULL IN (empty list) returns 0
+	if leftNull {
+		if values.Size() == 0 {
+			return false, nil
+		}
+		return nil, nil
+	}
+
+	// convert left to right's type
+	nLeft, _, err := typ.Convert(left)
+	if err != nil {
+		return false, nil
+	}
+
+	key, err := sql.HashOf(sql.NewRow(nLeft))
+	if err != nil {
+		return nil, err
+	}
+
+	val, notFoundErr := values.Get(key)
+	if notFoundErr != nil {
+		if _, nilValNotFoundErr := values.Get(nilKey); nilValNotFoundErr == nil {
+			return nil, nil
+		}
+		return false, nil
+	}
+
+	var r sql.Row
+	rowVal, ok := val.([]any)
+	if !ok {
+		r = sql.Row{val}
+	} else {
+		r = sql.NewRow(rowVal...)
+	}
+
+	return in.valuesEqual(ctx, left, r)
+}
+
+func (in *InSubquery) valuesEqual(ctx *sql.Context, left interface{}, row sql.Row) (interface{}, error) {
```

Comment:
Even more comments! Fairly hard to over-comment, but very easy to under-comment.